python/revng/internal/api/rich_manager.py

```diff
@@ -0,0 +1,226 @@
+
+            try:
+                self.synchronizer.save(last_dir, old_directories)
+            except self.synchronizer.save_exceptions() as e:
+                sys.stderr.write("Failed syncing resume directory\n")
+                traceback.print_exception(e, file=sys.stderr)
+                sys.stderr.flush()
+
+            for element in old_dir_list:
+                rmtree(element, False)
+
+            with self.lock:
+                credentials = self.credentials
+
+            for hook in self.save_hooks:
+                hook(self.manager, credentials)
+
+        self.shutdown_event.set()
+
+    def submit(self, dir_: str):
+        assert not self.stopping, "Cannot submit while stopping"
+        with self.lock:
+            self.dir_list.append(dir_)
+            self.submit_event.set()
+
+    def set_credentials(self, credentials: str):
+        with self.lock:
+            self.credentials = credentials
+            self.synchronizer.set_credentials(credentials)
+
+
+class ResettableTimer(Thread):
+    def __init__(self, interval: int, func: Callable[[], Any]):
+        super().__init__()
+        self.interval = interval
+        self.func = func
+        self.running = True
+
+        self.lock = Lock()
+        self.next_trigger_time = None
+
+    def run(self):
+        while self.running:
+            if self.next_trigger_time is not None and time.time() > self.next_trigger_time:
+                with self.lock:
+                    self.func()
+                    self.next_trigger_time = None
+            time.sleep(10)
+
+    def reset_countdown(self):
+        with self.lock:
+            self.next_trigger_time = None
+
+    def start_countdown(self):
+        with self.lock:
+            if self.next_trigger_time is None:
+                self.next_trigger_time = time.time() + self.interval
+
+
+class EventType(enum.StrEnum):
+    VOLATILE_CHANGE = enum.auto()
+    SAVE = enum.auto()
+
+
+class RichManager(Manager):
+    SAVE_INTERVAL = 120
+
+    def __init__(
+        self,
+        workdir: str | None = None,
+        flags: Iterable[str] = (),
+        save_hooks: SaveHooks = (),
+    ):
+        synchronizer: Synchronizer | None = None
+        self._storage_saver: StorageSaver | None = None
+
+        if workdir is None:
+            self._temporary_workdir = TemporaryDirectory(prefix="revng-manager-workdir-")
+            workdir = self._temporary_workdir.name
+        elif workdir.startswith("s3://") or workdir.startswith("s3s://"):
+            synchronizer = S3Synchronizer(workdir)
+            self._storage_saver = StorageSaver(self, synchronizer, save_hooks)
+            self._storage_saver.start()
+
+            self._temporary_workdir = TemporaryDirectory(prefix="revng-manager-workdir-")
+            workdir = self._temporary_workdir.name
+
+        if synchronizer is not None:
+            synchronizer.load(Path(workdir))
+
+        self.save_lock = Lock()
+        super().__init__(workdir, flags)
```

Comment:
I'd outline everything above this line to a function `get_workdir` and ensure `super().__init__` is called in the first line of `__init__`.

Reply:
Calling `__init__` on the first line is not a strict requirement, we're outlining a piece of code that's only called in `__init__` which adds class attributes (which makes linters unhappy) just so `__init__` can be on the first line. I'd argue to keep the code there.

Reply:
> I understand it's not a strict requirement, but, unless you're doing something to prepare an argument for the base class constructor, it's better to put it afterwards.

I'm not sure I understand the meaning of this sentence, we're preparing arguments for the parent's constructor, it needs to show up above `super().__init__`

Reply:
Yes, in fact that part is fine, but I'd outline it in a function.

This way, next time someone makes a change is not tempted to add stuff before `super().__init__()`, given that the first line will be `super().__init__(get_workdir(), flags)`.